Web app Merge Review

p1oab opened the pull request.

Commit:
Cleaned up a bunch of stuff in GPS.Data Added the two controllers.

Commit:
Added the code from GPS.Service to our Web API, deleted GPS.Service, then renamed our web app to GPS.Service.

Commit:
Finishing everything the previous commit said it would do.

Commit:
Changed GPS.Data to standard2.0

braydenperry committed:
Deleted references to GPS.Data.old.csproj

braydenperry committed:
Created basic API controller

braydenperry committed:
Rolled start and end date properties into DateTime properties

braydenperry committed:
Merge branch 'DateTime' into WebApp

braydenperry committed:
Merge branch 'API' into WebApp

braydenperry committed:
Revert "Merge branch 'API' into WebApp"

This reverts commit a2ae3a349cfbb2649c47d52ab519a1213431fb1f, reversing
changes made to cc800646fb348623ba028acb4f62b933aa6f75c1.

braydenperry committed:
Revert "Merge branch 'DateTime' into WebApp"

This reverts commit cc800646fb348623ba028acb4f62b933aa6f75c1, reversing
changes made to 1f6552625b122493fcc857b1ad10be78b1a56772.

braydenperry committed:
Added API functionality

braydenperry committed:
Rolled start and end attributes into DateTime attributes

braydenperry committed:
Rolled start and end attributes into DateTime attributes

Commit:
Wrote the outline for validation tests, added try catch blocks in parser.cs, and changed every variable type that was an int to a string. That was done in order to skip a tag if its attributes aren't valid instead of throwing an InvalidCastException error and shutting the program down.

Commit:
Added validation for isNaN and isNumber in validation.cs

tannerp6260 committed:
Added a basic paginated table to index.cshtml

tannerp6260 committed:
Added a basic, preliminary paginated table

Commit:
Validat.cs

Commit:
Merging Validation changes to the webapp branch.

braydenperry committed:
Upload via endpoints working.

hannahmjex committed:
Outage - now includes startTime and endTime again.

Parser should test - commented out all tests that include start/end year, doy, hr, min, sec. This test will need to be updated ASAP

Parser - added int.Parse to all start/end year, doy, hr, min, sec being used to set the StartTime and EndTime.

Historical - added a StartTime and EndTime attribute to be used in validation

Validate - created one test for historical to validate the endTime.

hannahmjex committed:
refined previous changes

hannahmjex committed:
Add date time and end time

hannahmjex committed:
Validation written for each outage type.
Added end time for predicted outage.

Commit:
Fixed the tests

Commit:
Fixed the tests to accomodate the DateTime variables.

braydenperry committed:
Rearranged some code

braydenperry committed:
Merge branch 'FrontEnd' into WebApp

hannahmjex committed:
Added tests to check start time and end time for all tags.

Added comments to explain tests

braydenperry committed:
Update Index.cshtml

hannahmjex committed:
merge validation continued to WebApp

hannahmjex committed:
Merge branch 'WebApp' of https://github.com/braydenperry/GPS into WebApp

braydenperry committed:
Merge branch 'Upload' into WebApp

braydenperry committed:
Delete Endpoint working

braydenperry committed:
Centered table and made it (more) responsive.

braydenperry committed:
Fixed visual glitch (table overlapping footer).

braydenperry committed:
Button working.

tannerp6260 committed:
Moved closing tags for 'tbody' and 'tfoot' to correct locations; removed unnecessary text from page; changed table tfoot text to match thead text

hannahmjex committed:
Commented out all SOF files in GPS.Data to see if they were used for populating the list. They aren't. :)

hannahmjex committed:
Removed SOF folder and all files from GPS.Data. They were not being used.

hannahmjex committed:
Updated test dependencies to include GPS.Service instead of GPS.Data, because the SOF files live inside GPS.Service.

tannerp6260 committed:
Fixed issues for table centering, blank cell in table footer, and buttons being unclickable at bottom of page

Commit:
The list of errors in parser.cs should be in place. Now I need to get that outputting through api and a controller.

hannahmjex committed:
Changed page title to GPS Satellite Outages

Added a page in the navbar for Current Outages, Historical Outages, Predicted Outages, and SOF Download

braydenperry committed:
Merge branch 'TableCentering' into WebApp

Commit:
Got rid of the errorlog.txt files as well as the TODO for logging errors in validation. I'm doing it in parser instead.

hannahmjex committed:
Naviagtion bar includes an "All Outages" and a "SOF File" tab.

hannahmjex committed:
Merge branch 'CombineSOF' into WebApp

hannahmjex committed:
Added parser should not test file

hannahmjex committed:
Added ParserShouldNot tests

hannahmjex committed:
Attempting to create the SOF file download page

braydenperry committed:
Made upload function more aesthetic

Commit:
Everything is working! Errors are entered into a list in parser.cs and outputted through a controller at .../api/ViewErrors

braydenperry committed:
Fixed upload

hannahmjex committed:
Added a razor page for the SOFDownload link.

Added a folder in wwwroot that contains the current SOF file to be downloaded.

Commit:
Merge branch 'LogParserErrors' into WebApp

Commit:
I made very minimal changes, but the lock around the serializer in parser is there. I don't think we will need a lock anywhere else, but am unsure.

hannahmjex committed:
Merged EditNavigationBar to WebApp

hannahmjex committed:
Merge branch 'WebApp' of https://github.com/braydenperry/GPS into WebApp

hannahmjex committed:
Changed GPS.Test.csproj

hannahmjex committed:
When merging from EditNavigationBar, the table was repeated. It has now been deleted and webapp runs

tannerp6260 committed:
Eddited data-table.js during merge.

tannerp6260 committed:
Merge branch 'WebApp' of https://github.com/braydenperry/GPS into WebApp

tannerp6260 committed:
Merged FrontEnd_Formatting to WebApp

Commit:
Merge branch 'WebApp' of https://github.com/braydenperry/GPS into WebApp

hannahmjex committed:
Merged Buttons to Webapp

hannahmjex committed:
Merged Buttons to WebApp

braydenperry committed:
Merge branch 'WebApp' into Buttons

Commit:
Deleted a test OnGet function in cshtml that wasn't being used anymore.

Commit:
Merge branch 'WebApp' of https://github.com/braydenperry/GPS into WebApp

braydenperry committed:
Validates that the uploaded file has .sof extension. Deletes the SOF file if it already exists when a new one is uploaded.

p1oab commented on GPS.Data/Parser.cs:
```diff
@@ -1,101 +1,160 @@
-        {
-            //Serializes the .sof and populates each of the classes
-            Serializer = new XmlSerializer(typeof(GpsIsFile));
-            using Stream reader = new FileStream(filePath, FileMode.Open);
-            Outages = (GpsIsFile)Serializer.Deserialize(reader);
```
Cannot rely on using statement to dispose if exception can be thrown from object other than the object in the using statement (eg, reader will only be disposed on an exception if reader itself throws an exception,  If Serializer.Deserialize can throw an exception reader.Dispose will not be called).  If this is the case, use try/finally pattern with the reader.Dispose called in the finally.  Should be applied to all situations

hannahmjex replied:
declare reader outside try as null
create it inside a try
null check
finally

hannahmjex replied:
Don't use using anymore

p1oab commented on GPS.Data/Parser.cs:
```diff
@@ -1,101 +1,160 @@
+			{
+				Console.WriteLine("FileNotFoundException");
+				//TODO: navigate to an error page if this is the case
+				System.Environment.Exit(1);
```
This will be a class used outside of test program.  Allowing it to throw an exception is good enough (used exit at the top level where you within the application itself)

p1oab commented on GPS.Data/Parser.cs:
```diff
@@ -1,101 +1,160 @@
-    }
+			if (_errorLog.Count == 0)
+			{
+				_errorLog.Add("There were no errors with the current.sof file");
```
Leave error log empty if no errors

connorwood2 replied:
can move into webpage

braydenperry committed:
Comments and code cleanup.

p1oab commented on GPS.Data/Validate.cs:
```diff
@@ -0,0 +1,176 @@
+
+namespace GPS.Data
+{
+	public class Validate
```
This would be a good candidate to make extension methods since there is no instance members here.  This way you can just call .Validate off the existing object

p1oab commented on GPS.Data/Validate.cs:
```diff
@@ -0,0 +1,176 @@
+		}
+
+		/// <summary>
+		/// Returns 1 if valid and end time exists
+		/// Returns 2 if valid and NO end time
+		/// Returns 3 if invalid
```
Instead of using int, make an enum to represent your validation status

p1oab commented on GPS.Service/Conrollers/ManageOutagesController.cs:
```diff
@@ -0,0 +1,79 @@
+            {
+                if (file.File.Length > 0) // Make sure there's actually a file
+                {
+                    using FileStream fileStream = System.IO.File.Create(_environment.WebRootPath + "\\SOF\\" + file.File.FileName);
```
Should create an interface to access the outage repository (IOutageRepository or something like that).  Later if we change to some other db, then none of the code will need to be modified outside of creating a new implementation of IOutageRepository.  (should have Import, Delete methods on it)

connorwood2 replied:
Brayden has more notes on this

hannahmjex replied:
In data service, each controller is independent and have access to same repo object

p1oab commented on GPS.Service/Conrollers/ManageOutagesController.cs:
```diff
@@ -0,0 +1,79 @@
+                    using FileStream fileStream = System.IO.File.Create(_environment.WebRootPath + "\\SOF\\" + file.File.FileName);
+                    await file.File.CopyToAsync(fileStream);
+                    fileStream.Flush();
+                    return "File uploaded.";
```
Should be validating the file before overwriting it (this can happen in the repository import method itself)

p1oab commented on GPS.Service/Conrollers/ManageOutagesController.cs:
```diff
@@ -0,0 +1,79 @@
+        }
+        
+        [HttpPost]
+        public async Task<string> Post([FromForm] FileUpload file)
```
In your controller methods, you should return Task<IActionResult>, that way you can take advantage of the build in macros for error codes:  Ok, BadRequest, NotFound, etc.  Rather than just a message, we want to return common http request error codes (they may additionally have other data returned including a message)

p1oab commented on GPS.Service/Conrollers/QueryOutagesController.cs:
```diff
@@ -0,0 +1,37 @@
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class QueryOutagesController : Controller
```
Might not have specified, but we want the backend data service to be a separate service from the UI web app.  There will be other consumers of the data besides the UI frontend.  The UI would access the data via http requests

braydenperry committed:
Merge branch 'Buttons' into WebApp

tannerp6260 committed:
Changed a comment that was inaccurate.

tannerp6260 committed:
Decreased .mydatatable tfoot input 'max-width' to 120px to fix table centering/sizing issue.

Commit:
Greated a new DataService project and moved the controllers from our old GPS.Serive project into the new project. We also moved the SOF folder from GPS.Service to GPS.Data and overloaded the constructor for Parser.cs.

braydenperry committed:
Controllers and Repository implementation

Commit:
GPS.Service is now pulling from GPS.DataService. GPS.Data, Tests, and DataService should know nothing about GPS.Service at this point.

tannerp6260 committed:
Removed an old comment

tannerp6260 committed:
Modified bottom row of table to use dropdown search for 'Tag Name' and 'Type' columns and text search for all other columns.

tannerp6260 committed:
Removed commented segment of code.

tannerp6260 committed:
Changed max-width of input to 100% instead of using a static value

tannerp6260 committed:
At the bottom of each column I changed the dropdown menu (select) min-height to 30px to match the height of the search box.

tannerp6260 committed:
Removed unused variable

tannerp6260 committed:
Modified 'th' parameters in site.css to properly scale datatable and fit within container (fixes centering issue).

tannerp6260 committed:
Implemented 'ultimate date/time sorting' plugin. Now the dates sort properly!

tannerp6260 committed:
comment typo

tannerp6260 committed:
added url reference to date sorting plugin

tannerp6260 committed:
Modified script to be contained entirely within '.ready()' function. This fixxed a locally scoped variable issue causing search filters to stop working.

braydenperry committed:
Upload working now

tannerp6260 committed:
Changed .mydatatable max-width from 50px to 70px to fix problem of sorting arrows overlapping 'th' text when scaling page.

braydenperry committed:
Exception handler changed to be more generic.

braydenperry committed:
Refactored data table to call API instead of repository (the GPS.Service project shouldn't know about the repository)

braydenperry committed:
Refactored the data tables

Commit:
Added the Enum for Validate.cs

Commit:
Added and enum to parser.cs

Commit:
Made the enum private.

tannerp6260 committed:
Began localizing web references for datatable

braydenperry committed:
Moved SOF folder into a single place (GPS.Data)

braydenperry committed:
Merge branch 'RefactorEverything' of https://github.com/braydenperry/GPS into RefactorEverything

tannerp6260 committed:
localized all href and src online references

tannerp6260 committed:
added dataTables.bootstrap4.min.css to directory

tannerp6260 committed:
Completed localizing src and href online references in _layout.cshtml. Added respective js and css files to repository

tannerp6260 committed:
Revert "Changed .mydatatable max-width from 50px to 70px to fix problem of sorting arrows overlapping 'th' text when scaling page."

This reverts commit 79f36379e30db0ddeb656e5ea2105cf49ad452ba.

tannerp6260 committed:
Changed an incorrect revert. To do this I changed .mytable max-width from 50px to 70px

tannerp6260 committed:
Fixed typo

tannerp6260 committed:
Revert "localized all href and src online references"

This reverts commit 3789a5957465c702e964f7164eb3bd9fe9048113.

tannerp6260 committed:
Merge pull request #13 from braydenperry/SearchFunctionality

Search functionality

tannerp6260 committed:
Removed 'smart search' option from data-table.js. I accidentally included this in a previous commit that I did not mean to.

Commit:
Nothing happened

Commit:
Almost all of the notes from our meeting with Arnold have been completed in the refactor everything branch. That branch is what is being merged into our "develop" or WebApp branch.

Commit:
Added Status Code return error for Query Outages. I don't know how to get that to display through the javascript/AJAX code, though.

Also, we (Hannah and I) tried to fix the tests on this branch by adding some more code to the overloaded constructor in parser.cs, but that didn't work.

ALSO!!!!! SOMEONE DELETED OUR INVALID.SOF FILE AND THAT'S A PROBLEM! >:/

Commit:
Tests are working again!

tannerp6260 committed:
Began frontend refactoring by implementing date-sorting function (moment) in brayden's provided data-table.js code.

tannerp6260 committed:
Implemented old 'moment' function in a way that works. Brayden's example code is not working yet.

Commit:
Fixing invalidTest.sof

tannerp6260 committed:
Utilized previous front end data-table.js code to implement dropdown menus and regex searching. The dropdown tables are not populating with options though.

Commit:
The invalidTest.sof file is back to where it needs to be. I am still unsure if I am returning errors in my controllers correctly.

tannerp6260 committed:
Reorganized code for clarity

tannerp6260 committed:
Changed variable names for clarity

braydenperry committed:
Creation and Reference tags added

braydenperry committed:
Merged RefactorEverything into WebApp

braydenperry committed:
Finialized remaining merge conflicts

Commit:
Fixed the tests a little. Creating a new branch to fix the rest.

Commit:
Fixed the tests and added code to validate creation and reference tags.

p1oab commented on GPS.Data/Parser.cs:
```diff
@@ -12,14 +14,90 @@ public class Parser
+        /// <summary>
+        /// Constructor that is hard-coded to route to the current.sof file.
+        /// </summary>
+        public Parser()
```
easier to work with to just have the full path to the file passed into the class in constructor at all times.
you would just need to move that code that gets solution path to the class that is calling this.

Since we could be running this locally or from a container the basepath needs to be easily modifiable

p1oab commented on GPS.Data/Repository/IOutageRepository.cs:
```diff
@@ -0,0 +1,18 @@
+    public interface IOutageRepository
+    {
+        public List<Outage> Get();
+
```
For this interface, Have Stream for the type.  This allows it to be more generic than specifically an IFormFile.
You can get the stream from the IFormFile when you utilize the interface

p1oab commented on GPS.Data/Repository/OutageRepository.cs:
```diff
@@ -0,0 +1,114 @@
+
+        public OutageRepository()
+        {
+            
```
_allOutages = parser.populate.... is called twice.  only need to call once (this should be in the lock)

p1oab commented on GPS.Data/Repository/OutageRepository.cs:
```diff
@@ -0,0 +1,114 @@
+
+                if (file.Length > 0) // Make sure there's actually a file being uploaded.
+                {
+
```
Need to attempt to open the file in the parser.  If its readable returns some outages, then you can overwrite it.  Otherwise, should return error code and leave existing file alone

p1oab commented on GPS.Data/Repository/OutageRepository.cs:
```diff
@@ -0,0 +1,114 @@
+        public void Upload(IFormFile file)
+        {
+
+            lock (SOFFileLock)
```
only lock after the file has been read into the parser and you are ready to replace it.  You want to limit the amount of time that it spends locked

p1oab commented on GPS.Data/Repository/OutageRepository.cs:
```diff
@@ -0,0 +1,114 @@
+                        using FileStream fileStream = File.Create(_sofPath);
+                        file.CopyTo(fileStream);
+                        fileStream.Flush();
+                    }
```
need to update data (alloutages)

p1oab commented on GPS.Data/Repository/OutageRepository.cs:
```diff
@@ -0,0 +1,114 @@
+                {
+                    // If it does, delete it.
+                    File.Delete(_sofPath);
+                }
```
Need to update alloutages (clear it)

p1oab commented on GPS.Data/Validate.cs:
```diff
@@ -0,0 +1,228 @@
+		/// </summary>
+		/// <param name="outage"></param>
+		/// <returns></returns>
+		public int ValidatePredicted(Predicted outage)
```
Should be using the enum instead of the int here for the return type

p1oab commented on GPS.DataService/Controllers/QueryOutagesController.cs:
```diff
@@ -0,0 +1,49 @@
+            {
+                if (_outageRepository.Get() == null)
+                {
+                    throw new ArgumentException("There were no outages: Possible FileNotFound Exception");
```
Rather than relying on the catch, return the status code directly.  in your catch (unknown exceptions) you should probably return 500

p1oab commented on GPS.DataService/Controllers/ViewErrorsController.cs:
```diff
@@ -0,0 +1,35 @@
+﻿using System.Collections.Generic;
```
Don't need this.  If we need to log errors we can do it in the manage controller on import

p1oab commented on GPS.Service/Startup.cs:
```diff
@@ -24,6 +21,9 @@ public Startup(IConfiguration configuration)
         {
             services.AddRazorPages();
+            services.AddControllers();
+            // NOTE: FOR TESTING PURPOSES ONLY.
```
You will need to create another class that Implements IOutageRepository or you can make a new interface IOutageService with formats that better align (IFormFile in this case on the import).  Create a new class that will use HttpClient/Httpwebrequest to communicate with the data service

tannerp6260 committed:
Specified "Reference" column as initially sorted in ascending order per Conner's request.

tannerp6260 committed:
Modified code to correctly implement footer search and dropdown filters to run through the api.

braydenperry committed:
Merge remote-tracking branch 'origin/FrontEnd_Refactoring' into WebApp

braydenperry committed:
Merge remote-tracking branch 'origin/OfflineReferencing' into WebApp

tannerp6260 committed:
added offline references

tannerp6260 committed:
Added 2nd half of offline references
